server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -140,31 +142,47 @@ public void execute(SearchContext context) {
         }
 
         try {
-            SearchHit[] hits = new SearchHit[context.docIdsToLoadSize()];
-            FetchSubPhase.HitContext hitContext = new FetchSubPhase.HitContext();
+            // group docIds by segment in order to better use LRU cache
+            Map<Integer, List<Integer>> segmentTasks = new HashMap<>();
+            Map<Integer, Integer> docIdToIndex = new HashMap<>();
             for (int index = 0; index < context.docIdsToLoadSize(); index++) {
```

Comment:
We could sort the doc ids to load once and move the `LeafReaderContext` while iterating ? In fact that's what we do already in [FetchDocValuesPhase](https://github.com/elastic/elasticsearch/blob/954afd94fe859209522a6d63fcb15dd9ef6c9df2/server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchDocValuesPhase.java#L76) and other fetch sub-phases. Sorting the doc ids would remove the need to sort hits on every sub-phase.

Reply:
Thanks for your comment @jimczi .  Sort the docIdsToLoad looks better.

Reply:
Thanks for your comment @jimczi .  Sort the docIdsToLoad looks better. But [TopHitsAggregatorTests](https://github.com/elastic/elasticsearch/blob/master/server/src/test/java/org/elasticsearch/search/aggregations/metrics/TopHitsAggregatorTests.java#L106) would get failed if we change the order of docIdsToLoad.

Reply:
We also need to preserve the original order of hits in the response so sorting the doc ids is only internal (for fetching stored values and executing sub-phases). The original order must be restored when setting the hits in the [response](https://github.com/elastic/elasticsearch/pull/57273/files#diff-9c70930068120c515204efa560512fd9R200).

Reply:
Thanks for you comment @jimczi. Sort the doc ids to load seems better. But if we change the order of doc ids, the order of the search hits will relatedly be changed, it would cause a lot of test cases to fail, such as [testInsideTerms](https://github.com/elastic/elasticsearch/blob/master/server/src/test/java/org/elasticsearch/search/aggregations/metrics/TopHitsAggregatorTests.java#L106).

Reply:
> But if we change the order of doc ids, the order of the search hits will relatedly be changed, it would cause a lot of test cases to fail, such as testInsideTerms.

See my previous [comment](https://github.com/elastic/elasticsearch/pull/57273#discussion_r432285881).
We can change the order in the fetch phase but we have to preserve the original order in the response. The final [hits](https://github.com/elastic/elasticsearch/pull/57273/files#diff-9c70930068120c515204efa560512fd9R200) must be re-sorted based on their original order in the request (`context.docIdsToLoad`).